src/pages/PrivateDashboard.tsx

```diff
@@ -174,29 +190,221 @@ const SubmissionInspector: React.FC<{
     [inspect, toast],
   )
 
+  const [selectedCategory, setSelectedCategory] = useState<string>()
+  const [carouselSubmissions, setCarouselSubmissions] = useState<
+    SubmissionSsePayload[]
+  >([])
+  const LEFTOVER_CATEGORY_KEY = "none"
+
+  const handleCategorySelection = (categoryKey: string) => {
+    if (Object.keys(categories).length > 1)
+      setSelectedCategory((prev) =>
+        prev === categoryKey ? undefined : categoryKey,
+      )
+  }
+
+  const { categorize, isLoading } = useCategorize()
+
+  // Utils
+  const getSubmissionsAvgScore = (ids: number[]) => {
+    const selectedPoints = submissions
+      .filter((f) => ids.includes(f.submissionId))
+      .map((s) => s.points)
+
+    return selectedPoints.reduce((a, b) => a! + b!, 0)! / selectedPoints.length
+  }
+
+  const getSubmissionColor = (submissionId: number) => {
+    const foundColor = Object.values(categories).filter((category) =>
+      category.ids.includes(submissionId),
+    )
+
+    if (foundColor.length > 0) return foundColor[0].color
+    return "gray"
+  }
+
+  // Handle the fetching button
+  const handleFetchCategories = async () => {
+    categorize(submissions.map((s) => s.submissionId)).then((res) => {
+      const rawColors = ["purple", "green", "yellow", "blue", "orange"]
+      const newCategories: CategoriesType = {}
+
+      Object.keys(res.categories)
+        .sort((a, b) => res.categories[b].length - res.categories[a].length)
+        .map((key, i) => {
+          newCategories[`cat-${i}`] = {
+            color: rawColors[i],
+            ids: res.categories[key],
+            selectedIds: res.categories[key],
+            avgScore: getSubmissionsAvgScore(res.categories[key]),
+          }
+        })
+
+      const noCatIds = submissions
+        .map((s) => s.submissionId)
+        .filter((f) => !Object.values(res.categories).flat().includes(f))
+
+      if (noCatIds.length > 0) {
+        newCategories[LEFTOVER_CATEGORY_KEY] = {
+          color: "gray",
+          ids: noCatIds,
+          selectedIds: noCatIds,
+          avgScore: getSubmissionsAvgScore(noCatIds),
+        }
+      }
+
+      setCategories(newCategories)
+    })
+  }
+
+  // Handle the "None" category
+  useEffect(() => {
+    const withCatIds = Object.keys(categories)
+      .filter((f) => f !== LEFTOVER_CATEGORY_KEY)
+      .map((key) => categories[key].ids)
+      .flat()
+
+    const noCatIds = submissions
+      .map((s) => s.submissionId)
+      .filter((f) => !withCatIds.includes(f))
+
+    if (noCatIds.length > 0)
+      setCategories((prev) => ({
+        ...prev,
+        [LEFTOVER_CATEGORY_KEY]: {
+          color: "gray",
+          ids: noCatIds,
+          selectedIds: noCatIds,
+          avgScore: getSubmissionsAvgScore(noCatIds),
+        },
+      }))
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [submissions])
```

Comment:
If we put getSubmissionsAvgScore in a useCallback, I think we should be able to remove disabled es lint rule and include everything needed, should work as intended.